File: JavaScript/AoC2025/days/day12.js
import fs, { rmSync } from "fs"
import path from "path"

export default class Day12 {
    constructor() {
        const filePath = path.resolve("../../Data/2025/day12.txt");
        this.data = fs.readFileSync(filePath, "utf-8").split('\n').map(line => line.trim());
    } 

    printResults() {
        console.log("Day 12:");
        console.log(`Part 1: ${this.part1()}`);
        console.log(`Part 2: ${this.part2()}`);
    }

    part1() {
        const sizes = {};
        let num = 0;
        let present = 0;
        let region = 0;
        let actualSize = 0;
        let total = 0;
        for (const line of this.data) {
            if (line === '') {
                sizes[num] = present;
                present = 0;
            } else if (line.length === 2) {
                num = parseInt(line[0]);
            } else if (line.includes('#')) {
                for (const c of line) {
                    if (c === '#')
                        present++;
                }
            } else {
                const info = line.split(' ');
                region = parseInt(info[0].slice(0, 2)) * parseInt(info[0].slice(3, 5));

                for (let i = 1; i < info.length; i++) {
                    actualSize += sizes[i - 1] * parseInt(info[i]);
                }

                if (actualSize <= region)
                    total++;

                actualSize = 0;
            }
        }
        return total;
    }

    part2() {
        return "Marry Christmas!";
    }

}

File: CSharp/AoC2025/Days/Day12.cs
namespace AoC2025.Days;

public static class Day12
{
    // Data methods

    private readonly static string[] _data = File.ReadAllLines("../../Data/2025/day12.txt");

    // Solution

    public static int PartOne()
    {
        Dictionary<int, int> size = [];

        int num = 0, present = 0, region = 0, total = 0;
        foreach (var line in _data)
        {
            if (line == "")
            {
                size[num] = present;
                present = 0;
            }
            else if (line.Length == 2)
                num = int.Parse(line[0].ToString());
            else if (line.Contains('#'))
                present += line.Where(x => x == '#').Count();
            else
            {
                var nums = line.Split(' ');
                region = int.Parse(nums[0][..2]) * int.Parse(nums[0][3..5]);
                int space = 0;
                for (int i = 1; i < nums.Length; i++)
                {
                    space += int.Parse(nums[i]) * size[i - 1];
                }
                total += space <= region ? 1 : 0;
            }
        }
        return total;
    }

    public static string PartTwo()
    {
        return "Merry Christmas";
    }
}